hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/ConfiguredFailoverProxyProvider.java

```diff
@@ -93,4 +109,65 @@ public synchronized void close() throws IOException {
   public boolean useLogicalURI() {
     return true;
   }
+
+  /**
+   * Write active NameNode index to cache file.
+   */
+  public void writeActiveCache(int index) {
+    if (!cacheActiveEnabled) {
+      return;
+    }
+
+    synchronized (cacheActiveFile) {
+      boolean exist = cacheActiveFile.exists();
+      try (RandomAccessFile raf = new RandomAccessFile(cacheActiveFile, "rw");
+          FileChannel fc = raf.getChannel();
+          FileLock lock = fc.tryLock(0, Long.MAX_VALUE, false)) {
```

Comment:
I wasn't familiar with FileLock. Java's FileLock depends on the underlying OS implementation. Accordingly to https://stackoverflow.com/questions/12651068/release-of-flock-in-case-of-errors/12652718#12652718 Linux releases the lock even if the JVM crashes abnormally.

Reply:
Just adding the comment here in case other reviewers have the same doubt.

Reply:
Yes, according to my tests, after a process ends, the linux kernel releases all resources it holds, including locks.